Add Check Support for Java 21 Pattern Matching for Switch Syntax: NoWhitespaceBeforeCaseDefaultColon

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation:https://checkstyle.org/checks/whitespace/nowhitespacebeforecasedefaultcolon.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---
```

PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, boolean c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {   // NPath = 6
        switch (obj) {
            case ColoredPoint(int a, int b, _) when (a >= b) : {} break; // violation
            case ColoredPoint(int a, int b, _) when (b > 100) : {} break; // violation
            case ColoredPoint(int a, int b, _) when (b != 1000) : {} break; // violation
            case ColoredPoint(int a, int b, _) when (b != 100000) : {} break; // violation
            case Rectangle(_,_) : {} // violation
            default : System.out.println("none"); // violation
        };
    }
}
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="NoWhitespaceBeforeCaseDefaultColon">
        </module>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:7:62: ':' is preceded with whitespace. [NoWhitespaceBeforeCaseDefaultColon]
[ERROR] D:\CS\test\src\Test.java:8:63: ':' is preceded with whitespace. [NoWhitespaceBeforeCaseDefaultColon]
[ERROR] D:\CS\test\src\Test.java:9:65: ':' is preceded with whitespace. [NoWhitespaceBeforeCaseDefaultColon]
[ERROR] D:\CS\test\src\Test.java:10:67: ':' is preceded with whitespace. [NoWhitespaceBeforeCaseDefaultColon]
[ERROR] D:\CS\test\src\Test.java:11:33: ':' is preceded with whitespace. [NoWhitespaceBeforeCaseDefaultColon]
[ERROR] D:\CS\test\src\Test.java:12:21: ':' is preceded with whitespace. [NoWhitespaceBeforeCaseDefaultColon]
Audit done.
Checkstyle ends with 6 errors.
PS D:\CS\test> 

```

---

This check is working as expected.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheckTest.java
@@ -88,4 +88,20 @@ public class NoWhitespaceBeforeCaseDefaultColonCheckTest
                 .isEqualTo(check.getAcceptableTokens());
     }
 
+    @Test
+    public void testPatternMatchingForSwitch() throws Exception {
+        final String[] expected = {
+            "14:62: " + getCheckMessage(MSG_KEY, ":"),
+            "16:21: " + getCheckMessage(MSG_KEY, ":"),
+            "18:21: " + getCheckMessage(MSG_KEY, ":"),
+            "20:67: " + getCheckMessage(MSG_KEY, ":"),
+            "23:36: " + getCheckMessage(MSG_KEY, ":"),
+            "25:21: " + getCheckMessage(MSG_KEY, ":"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputNoWhitespaceBeforeCaseDefaultColonPatternMatchingForSwitch.java"),
+                expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColonPatternMatchingForSwitch.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespacebeforecasedefaultcolon/InputNoWhitespaceBeforeCaseDefaultColonPatternMatchingForSwitch.java
@@ -1,0 +1,42 @@
+/*
+NoWhitespaceBeforeCaseDefaultColon
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespacebeforecasedefaultcolon;
+
+public class InputNoWhitespaceBeforeCaseDefaultColonPatternMatchingForSwitch {
+
+    void test(Object obj) {
+        switch (obj) {
+            case ColoredPoint(int a, int b, _) when (a >= b) : {} break; // violation
+            case ColoredPoint(int a, int b, _) when (b > 100)
+                    : {} break; // violation
+            case ColoredPoint(int a, int b, _) when (b != 1000)
+                    :  // violation
+            {} break;
+            case ColoredPoint(int a, int b, _) when (b != 100000) : // violation
+            {} break;
+            case
+                    Rectangle(_,_) :  // violation
+            {}
+            default : System.out.println("none"); // violation
+        }
+    }
+
+    void test2(Object obj) {
+        switch (obj) {
+            case ColoredPoint(int a, int b, _) when (a >= b): {} break;
+            case ColoredPoint(int a, int b, _) when (b > 100): {} break;
+            case ColoredPoint(int a, int b, _) when (b != 1000): {} break;
+            case ColoredPoint(int a, int b, _) when (b != 100000): {} break;
+            case Rectangle(_, _): {}
+            default: System.out.println("none");
+        }
+    }
+
+    record ColoredPoint(int p, int x, boolean c) { }
+    record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
+}

